tracing/hooks.go

```diff
@@ -3,17 +3,35 @@ package tracing
 // BaseplateHook allows you to inject functionality into the lifecycle of a
 // Baseplate request.
 type BaseplateHook interface {
-	OnServerSpanCreate(span *Span) error
+	// OnCreateServerSpan is called after a ServerSpan is first created by
+	// tracing.CreateServerSpan, before any OnStart Hooks are called.
+	//
+	// OnCreateServerSpan is the recommended place to register Hooks onto the
+	// ServerSpan.
+	OnCreateServerSpan(span *ServerSpan) error
 }
 
 // SpanHook allows you to inject functionality into the lifecycle of Baseplate
 // Spans.
 type SpanHook interface {
-	OnCreateChild(child *Span) error
-	OnStart(span *Span) error
-	OnEnd(span *Span, err error) error
-	OnSetTag(span *Span, key string, value interface{}) error
-	OnAddCounter(span *Span, key string, delta float64) error
+	// OnCreateChild is called after a child Span is first created, before any
+	// OnStart Hooks are called.
+	OnCreateChild(child Span) error
+
+	// OnStart is called after a child Span is created and the OnCreateChild
+	// Hooks on the parent Span are called.
+	OnStart(span Span) error
+
+	// OnStop is called by Span.Stop, after calling span.PreStop but before
+	// serializing and publishing the Span.
+	OnStop(span Span, err error) error
```

Comment:
🔕 Should we update `trace.stop` in `PreStop` based on this description? But I think the hook implementation getting span in this function has no real access to the stop time so it probably doesn't matter.

Reply:
Ya, I think we want to leave `trace.stop` out of pre-stop since it's something we always want to update.

Reply:
I think another point is, do we want to count the time hooks took into the span time? that can also affect when we record the stop (and start) time

Reply:
Based on the answer to that question, we probably want to rename the hook functions to be more clear (e.g. `OnPreStop` if we count the hook times, or `OnPostStop` if we don't).

Reply:
🤔 

I think we do?  One issue we've ran into in the past is losing time data around things that happen outside of spans (like thrift serialization/deserialization time), so, as you mentioned in another place, calling this `OnPreStop` and running it before truly "stopping" might make sense.